File: eliHenderson/assignments/BreadthFirstSearch/lc111.js
// Find the minimum depth of a binary tree. The minimum depth is the number of nodes along the shortest path from the root node to the nearest leaf node.
/**
 *          12
 *        7    1
 *       9   10 5
 *          11
 *
 * Overview:
 *  Follow the BFS pattern
 *  set a new queue
 *  insert the root into the queue
 *  set a minDepth counter = 0
 *  while the queue is not empty
 *  loop through the queue
 *      store the queue size as level size
 *      for loop from 0 - level size
 *      store the current node
 *      check if the current level has a leaf, if so return min depth counter
 *      check for children
 *      if there are children, insert them into the queue
 *      leave the for loop
 *   increment the minDepth counter
 *
 * return max depth counter
 *
 *
 *  sudo:
 *      let queue = [root]
 *      let minDepth = 0
 *      while(queue.len)
 *          lvlSize = queue.len
 *          for( 0 - lvlSize)
 *          currNode = queue.shift
 *          if(!currNode.left && !currNode.right) return minDepth
 *          if(currNode.left) queue.push(currNode.left)
 *          if(currNode.right) queue.push(currNode.right)
 *
 *      minDepth++
 *
 *
 * edge: if the root is null
 *          return 0
 *     root value is null
 *          return 1
 *          because a node of value null is still valid
 *
 *
 * Time: O(N)
 * Space: O(N)
 *
 */

class TreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

const find_minimum_depth = function (root) {
  if (root === null) return 0;
  //create the queue
  let queue = [];
  //insert the root
  queue.push(root);
  //create a minDepth counter
  let minDepth = 0;
  //loop while queue is not empty
  while (queue.length) {
    //increment
    minDepth++;
    let lvlSize = queue.length;
    //loop through the level
    for (let i = 0; i < lvlSize; i++) {
      //hold currentNode
      let currentNode = queue.shift();

      //check if we are at leaf
      //if so, return minDepth
      if (!currentNode.left && !currentNode.right) return minDepth;

      //check for children
      if (currentNode.left) queue.push(currentNode.left);
      if (currentNode.right) queue.push(currentNode.right);
    }
  }
};

const find_minimum_depth = (root) => {
  if (root === null) return 0;
  //create the queue
  let queue = [];
  //insert the root
  queue.push(root);
  //create a minDeptCounter;
  let minDepth = 0;
  //loop while queue is not empty
  while (queue.length) {
    //increment
    minDepth++;
    //we must increment before going back into the loop
    //this way, we don't return before counting the current level
    let levelSize = queue.length;
    for (let i = 0; i < levelSize; i++) {
      //hold current Node
      let currentNode = queue.shift();

      //check if we are at leaf
      //if the current node is a leaf
      //return minDepth counter
      if (!currentNode.left && !currentNode.right) return minDepth;

      //check for children
      if (currentNode.left) queue.push(currentNode.left);
      if (currentNode.right) queue.push(currentNode.right);
    }
  }
};

var root = new TreeNode(12);
root.left = new TreeNode(7);
root.right = new TreeNode(1);
root.right.left = new TreeNode(10);
root.right.right = new TreeNode(5);
console.log(`Tree Minimum Depth: ${find_minimum_depth(root)}`);
root.left.left = new TreeNode(9);
root.right.left.left = new TreeNode(11);
console.log(`Tree Minimum Depth: ${find_minimum_depth(root)}`);

root = new TreeNode(null);
console.log(`Tree Minimum Depth: ${find_minimum_depth(root)}`);

File: mathewDavidov/assignments/treesBFS/lc111/lc111.java
/**
 * Problem Statement:
 * Find the minimum depth of a binary tree. 
 * The minimum depth is the number of nodes along the shortest path from the root node to the nearest leaf node.
 * 
 * Example:
 * Input: root = [3,9,20,null,null,15,7]
 *      3
 *     9 20
 *      15 7
 * Output: 2
 * 
 * Edge cases:
 * null root -> return 0
 * 
 * initialize a queue and add the root
 * initialize a min depth to 1
 * while the queue is not empty:
 *   initialize a variable to the current size of the queue
 *   loop from 0 to the size of the queue:
 *     poll the current element in the front of the queue
 *     if the current node's left and right child are null, return the min depth
 *     if the left child exists, add it to the queue
 *     if the right child exists, add it to the queue
 *   increment min depth by 1
 * return min depth
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(n)
 */
import java.util.*;

public class lc111 {
    public static int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);

        int minDepth = 1;

        while (!queue.isEmpty()) {
            int queueSize = queue.size();

            for (int i = 0; i < queueSize; i++) {
                TreeNode current = queue.poll();

                if (current.left == null && current.right == null) {
                    return minDepth;
                }

                if (current.left != null) {
                    queue.offer(current.left);
                }

                if (current.right != null) {
                    queue.offer(current.right);
                }
            }

            minDepth++;
        }

        return minDepth;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(12);
        root.left = new TreeNode(7);
        root.right = new TreeNode(1);
        root.right.left = new TreeNode(10);
        root.right.right = new TreeNode(5);
        System.out.println("Tree Minimum Depth: " + minDepth(root));
        root.left.left = new TreeNode(9);
        root.right.left.left = new TreeNode(11);
        System.out.println("Tree Minimum Depth: " + minDepth(root));
    }
}

class TreeNode {
    public int val = 0;
    public TreeNode left;
    public TreeNode right;
   
    public TreeNode(int val) {
       this.val = val;
       this.left = this.right = null;
    }
}